### PR TITLE
fix: Yet another attempt at preventing `Task already stopped` errors

### PR DESCRIPTION
**Pull request by PhilippeWeidmann**

### CODE REVIEW

**Review by valentinperignon — DISMISSED** [review]
Yet another approval

File: MailCore/Utils/URLSchemeHandler.swift [hunk 1]
```diff
@@ -39,27 +40,33 @@ public class URLSchemeHandler: NSObject, WKURLSchemeHandler {
...
-                urlSchemeTask.didFinish()
-            }
-            if let error {
-                urlSchemeTask.didFailWithError(error)
+                self?.dataTasksInProgress[urlSchemeTask.hash] = nil
+
+                if let response {
+                    urlSchemeTask.didReceive(response)
+                }
+                if let data {
```
Comment (adrien-coye) [inline MailCore/Utils/URLSchemeHandler.swift:53]: I would use a `else if`

File: MailCore/Utils/URLSchemeHandler.swift [hunk 2]
```diff
@@ -39,27 +40,33 @@ public class URLSchemeHandler: NSObject, WKURLSchemeHandler {
...
+                    urlSchemeTask.didReceive(data)
+                    urlSchemeTask.didFinish()
+                }
+                if let error {
```
Comment (adrien-coye) [inline MailCore/Utils/URLSchemeHandler.swift:57]: I would `if let error` first and return ASAP in error case.

File: MailCore/Utils/URLSchemeHandler.swift [hunk 3]
```diff
@@ -39,27 +40,33 @@ public class URLSchemeHandler: NSObject, WKURLSchemeHandler {
...
         dataTasksInProgress[urlSchemeTask.hash] = dataTask
         dataTask.resume()
     }
 
     public func webView(_ webView: WKWebView, stop urlSchemeTask: WKURLSchemeTask) {
-        let dataTask = dataTasksInProgress[urlSchemeTask.hash]
-        dataTasksInProgress[urlSchemeTask.hash] = nil
-        dataTask?.cancel()
+        syncQueue.sync {
+            dataTasksInProgress[urlSchemeTask.hash]?.cancel()
```
Comment (adrien-coye) [inline MailCore/Utils/URLSchemeHandler.swift:68]: I would `guard let task = dataTasksInProgress[urlSchemeTask.hash]`  first.

**Review by adrien-coye — COMMENTED** [review]
I made some comments.

File: MailCore/Utils/URLSchemeHandler.swift (hunk 1 above)
Comment (adrien-coye) [reply MailCore/Utils/URLSchemeHandler.swift:64]: discussed

**Review by adrien-coye — APPROVED** [review]
All good.

**Review by valentinperignon — APPROVED** [review]

**Pull request merged**